prow/jira/jira.go

```diff
@@ -259,8 +288,275 @@ func (jc *client) UpdateRemoteLink(id string, link *jira.RemoteLink) error {
+		if err != nil {
+			return nil, fmt.Errorf("error identifying unsettable fields during clone creation: %w", err)
+		}
+		createdIssue, err = jc.CreateIssue(childIssue)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// create clone links
+	link := &jira.IssueLink{
+		OutwardIssue: &jira.Issue{ID: parent.ID},
+		InwardIssue:  &jira.Issue{ID: createdIssue.ID},
+		Type: jira.IssueLinkType{
+			Name:    "Cloners",
+			Inward:  "is cloned by",
+			Outward: "clones",
+		},
+	}
+	if err := jc.CreateIssueLink(link); err != nil {
+		return nil, err
+	}
+	// Get updated issue, which would have issue links
+	if clonedIssue, err := jc.GetIssue(createdIssue.ID); err != nil {
+		// still return the originally created child issue here in case of failure to get updated issue
+		return createdIssue, fmt.Errorf("Could not get issue after creating issue links: %w", err)
+	} else {
+		return clonedIssue, nil
+	}
+}
+
+// CloneIssue copies an issue struct, clears unsettable fields, creates a new
+// issue using the updated struct, and then links the new struct as a clone to
+// the original.
+func (jc *client) CloneIssue(parent *jira.Issue) (*jira.Issue, error) {
+	return GenericCloneIssue(jc, parent)
+}
+
+func unsetProblematicFields(issue *jira.Issue, err error) (*jira.Issue, error) {
+	// handle unsettable "unknown" fields
+	// trim create error
+	jsonResponse := strings.TrimPrefix(err.Error(), "request failed. Please analyze the request body for more details. Status code: 400: ")
+	message := createIssueError{}
+	if newErr := json.Unmarshal([]byte(jsonResponse), &message); newErr != nil {
+		return nil, fmt.Errorf("Error: %v; Original jsonResponse: %v", newErr, err)
+	}
+	// turn issue into map to simplify unsetting process
+	marshalledIssue, err := json.Marshal(issue)
+	if err != nil {
+		return nil, err
+	}
+	issueMap := make(map[string]interface{})
+	if err := json.Unmarshal(marshalledIssue, &issueMap); err != nil {
+		return nil, err
+	}
+	fieldsMap := issueMap["fields"].(map[string]interface{})
+	for field := range message.Errors {
+		delete(fieldsMap, field)
+	}
+	issueMap["fields"] = fieldsMap
+	// turn back into jira.Issue type
+	marshalledFixedIssue, err := json.Marshal(issueMap)
+	if err != nil {
+		return nil, err
+	}
+	newIssue := jira.Issue{}
+	if err := json.Unmarshal(marshalledFixedIssue, &newIssue); err != nil {
+		return nil, err
+	}
+	return &newIssue, nil
+}
+
+type createIssueError struct {
+	ErrorMessages []string          `json:"errorMessages"`
+	Errors        map[string]string `json:"errors"`
+}
+
+// GenericGetIssueSecurityLevel is a generic function for GetIssueSecurityLevel that can work with any
+// valid Client.
+func GenericGetIssueSecurityLevel(client Client, issue *jira.Issue) (*SecurityLevel, error) {
+	// TODO: Add field to the upstream go-jira package; if a security level exists, it is returned
+	// as part of the issue fields
+	securityField, ok := issue.Fields.Unknowns["security"]
+	if !ok {
```

Comment:
Isn't that an error? I would not expect a function that gives me a nil error to also give me a nil result. This is likely to lead to NPDs for consumers that expect this to be set when it happens to be unset

Reply:
It's not really an error. Just like any field in Jira, if the field is not returned, it gets set to `nil` and is handled as if it is not set. Most items in `Fields` are pointers for this reason. If an issue is just using the default security level, then it is unset. In the `Issue` struct, this would be represented as `nil`. If this function were to return an error, I would assume something else went wrong instead of the field simply being unset.

Reply:
Uhm okay, maybe at least mention in the godoc that it can be nil?
Would be great to get this upstream to avoid the need for any of this. Is there an issue or PR we could reference here?

Reply:
Uhm okay, maybe at least mention in the godoc that it can be nil?
Would be great to get this upstream to avoid the need for any of this. Is there an issue or PR we could reference here?

Reply:
I've updated the godoc to be clearer on that and added a link in the TODO comment of the function to an issue in go-jira requesting the `Security` field.